### ISSUE TITLE
Better logging on unrecoverable shard allocation failures

### ISSUE THREAD

**Issue by samcday** [elastic/elasticsearch#12412]
Allocation a shard is subject to some pretty complex logic. It's easy to get into a unresolvable state where shards cannot be allocated. In this case, it would be great to have `ERROR` level log messages showing up so I know what the heck is going on as an operator.

Here's a scenario I ran into tonight.

<WARNING - LONG BORING STORY AHEAD. SKIP TO THE BOTTOM FOR TL;DR>

---

I'm upgrading a cluster with 3 master nodes and 6 data nodes which lives in AWS from 1.6.0 to 1.6.1. We've configured our cluster with `cluster.routing.allocation.awareness.attributes` set to `aws_availability_zone`. We use Ansible to orchestrate rollouts. Ansible started going through each node, disabling cluster allocation + applying changes / upgrading ES + bringing ES back up and waiting for cluster state to go back to green before moving on to the next node.

Everything was going great, until after we bounced the third data node and realised that the cluster was refusing to go back into a green state. Investigation revealed that a shard was stuck in UNASSIGNED. No clue why, no logs, no pending recoveries / relocations. Just stuck in unassigned.

So, I tried to relocate the shard manually using the `_reroute` api:

```
{
    "commands" : [ 
        {
          "allocate" : {
              "index" : "<index with a stuck shard>", "shard" : 0, "node" : "i-0ab4dfc1"
          }
        }
    ]
}
```

... except that request was being denied. Why? Because at this point I had upgraded 3 of my data nodes to 1.6.1, but *all* of them were in us-west-1a. The other 3 data nodes in us-west-1c were 1.6.0. Turns out ES will refuse to relocate a shard to a node which is an older version (which makes perfect sense). So, the end result is the shard couldn't go anywhere in us-west-1a (because it's the same availability zone), but it also couldn't go anywhere in us-west-1c (because no nodes in that AZ were 1.6.1 yet).

I resolved the issue by setting a transient cluster setting to disable `cluster.routing.allocation.awareness.attributes`, completed the rollout, then reenabled that config.

---

Anyway, long story is long, just wanted to paint a picture of how easy it is to create a rather complex situation that would have been significantly easier to debug if I had simply seen a log message from ES saying "hey, I can't relocate this replica because *blah*" with the same helpful explanation output that `_reroute` gives you on bad relocation requests.

**Comment by clintongormley** [comment]
Hi @samcday 

I think the `explain` parameter on the reroute command will help you here:

> If the explain parameter is specified, a detailed explanation of why the commands could or could not be executed is returned.

That said, i think it'd be a nice improvement to just say "assign this shard" without having to specify a destination, and then be able to get the explanations of why assignment failed.

Logging is problematic because routing decisions are made all the time, and so can produce copious logs.

**Comment by samcday** [comment]
> Routing decisions are made all the time

Yeah, I see your point. I just wonder if it would be possible to maybe crank up the logging for a short period after mutative changes have happened? Things like cluster nodes leaving / joining, or cluster allocation being completely disabled for a brief period. I'm sure such a request is terrifying and subject to a dizzying number of race conditions and nightmares attributed to distributed consensus, but I figure I'd ask anyway :P`

**Comment by dakrone** [comment]
This is resolved with the cluster allocation explain API added in https://github.com/elastic/elasticsearch/pull/17305 (available in 5.0+)
